contracts/Curation/GIVUnipool.sol

```diff
@@ -0,0 +1,223 @@
+        lastUpdateTime = lastTimeRewardApplicable();
+        if (account != address(0)) {
+            rewards[account] = claimableStream(account);
+            userRewardPerTokenPaid[account] = rewardPerTokenStored;
+        }
+        _;
+    }
+
+    function __GIVUnipool_init(
+        address _tokenDistribution,
+        uint256 _duration
+    ) public initializer {
+        __Ownable_init();
+        __ERC20_init("GIVpower", "POW");
+        tokenDistro = IDistro(_tokenDistribution);
+        duration = _duration;
+        periodFinish = 0;
+        rewardRate = 0;
+    }
+
+    function lastTimeRewardApplicable() public view returns (uint256) {
+        return MathUpgradeable.min(getTimestamp(), periodFinish);
+    }
+
+    function rewardPerToken() public view returns (uint256) {
+        if (totalSupply() == 0) {
+            return rewardPerTokenStored;
+        }
+        return
+            rewardPerTokenStored.add(
+                lastTimeRewardApplicable()
+                    .sub(lastUpdateTime)
+                    .mul(rewardRate)
+                    .mul(1e18)
+                    .div(totalSupply())
+            );
+    }
+
+    /**
+     * Function to get the current timestamp from the block
+     */
+    function getTimestamp() public view virtual returns (uint256) {
+        return block.timestamp;
+    }
+
+    /**
+     * Function to get the amount of tokens is transferred in the claim tx
+     * @notice The difference between what this returns and what the claimableStream function returns
+     *  will be locked in TokenDistro to be streamed and released gradually
+     */
+    function earned(address account) external view returns (uint256) {
+        uint256 _totalEarned = claimableStream(account);
+        uint256 _tokenDistroReleasedTokens = tokenDistro.globallyClaimableAt(
+            getTimestamp()
+        );
+        uint256 _tokenDistroTotalTokens = tokenDistro.totalTokens();
+
+        return
+            (_totalEarned * _tokenDistroReleasedTokens) /
+            _tokenDistroTotalTokens;
+    }
+
+    // @dev This does the same thing the earned function of UnipoolTokenDistributor contract does.
+    // Returns the exact amount will be allocated on TokenDistro
+    function claimableStream(address account) public view returns (uint256) {
+        return
+            balanceOf(account)
+                .mul(rewardPerToken().sub(userRewardPerTokenPaid[account]))
+                .div(1e18)
+                .add(rewards[account]);
+    }
+
+    function stake(address user, uint256 amount)
+        internal
+        updateReward(user)
+    {
+        require(amount > 0, "Cannot stake 0");
+        _mint(user, amount);
+        emit Staked(user, amount);
+    }
+
+    function withdraw(address user, uint256 amount)
+        internal
+        updateReward(user)
+    {
+        require(amount > 0, "Cannot withdraw 0");
+        _burn(user, amount);
+        if (balanceOf(user) == 0) {
```

Comment:
I think this should be better `claimableStream(user)  != 0` or maybe nothing at all. I dont think it's intuitive that only the `getReward` function es called when the user unlocks all the locked tokens

Reply:
Yep, I think it's the right thing to do.